.adr/0016-closing-go-routines.md

```diff
@@ -0,0 +1,128 @@
+# 0016 Closing Go-routines
+
+## Status
+
+Review
+
+## Context
+
+In our codebase we have a few structs that use long-running go-routines to handle tasks asynchronously (often things like sending out messages/notifications). However we don't always consider how to stop and clean up after these go-routines, and we aren't always consistent on how we approach it. In some scenarios these long-running go-routines can continue running after `Close` has been called. This can easily introduce subtle race conditions.
+
+Here's a list of structs that spin up long-running go-routines.
+
+- [The RPC client](https://github.com/statechannels/go-nitro/blob/0b5fa37613363720c91c115c3de252a39b1b1f0a/rpc/client.go#L14)
+- [The RPC server](https://github.com/statechannels/go-nitro/blob/0b5fa37613363720c91c115c3de252a39b1b1f0a/rpc/server.go#L223)
+- [Eth Chain service](https://github.com/statechannels/go-nitro/blob/0b5fa37613363720c91c115c3de252a39b1b1f0a/client/engine/chainservice/eth_chainservice.go#L244)
+- [The API client](https://github.com/statechannels/go-nitro/blob/0b5fa37613363720c91c115c3de252a39b1b1f0a/client/client.go#L87)
+
+## Decision
+
+**Note:** I use the term struct as a shorthand for a struct with a long-running go-routine, like the examples above.
+
+When a struct's `Close` function is called, it should block and not return until:
+
+- all go-routines it owns stop executing.
+- any closeable resources it owns are closed.
+
+By enforcing these constraints a running go-routine can be guaranteed that it's parent struct is in a "running" state. This rules out a large class of race conditions and errors such as a go-routine attempting to use a parent struct's resource that has been closed.
+
+To enforce these constraints we should follow this pattern in a struct's `Close` function:
+
```

Comment:
>0a. Consider whether a goroutine is necessary. If the operation is likely to be very fast and cannot be blocked, it may be better to run it synchronously.

How about something like this?

0a. Consider whether a goroutine is necessary. The use of go-routines introduces the possibility of a variety of complex concurrency issues. If a go-routine is used it **must be managed in safe manner according to this ADR**.